hasher-matcher-actioner/hma-lite/hmalite/app.py

```diff
@@ -1,31 +1,64 @@
 # Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
 
+import csv
+import os
+import os.path
+import shutil
+
 from flask import Flask, render_template, request, jsonify, send_from_directory
+from hmalite.config import HmaLiteDevConfig, HmaLiteProdConfig
 from hmalite.matcher import matcher_api
-from threatexchange.signal_type import index, pdq_index
 from threatexchange.hashing.pdq_hasher import pdq_from_file
-import os
-import os.path
-import csv
+from threatexchange.signal_type import pdq_index
 
-CONFIG_ENV = "HMALITE_CONFIG_FILE"
 
 app = Flask(__name__)
 app.register_blueprint(matcher_api, url_prefix="/v1/hashes")
 
-
+# Set up those configs
 if app.config["ENV"] == "production":
-    app.config.from_object("hmalite.config.HmaLiteProdConfig")
+    config_cls = HmaLiteProdConfig
+    app.config.from_object(HmaLiteProdConfig())
 else:
-    app.config.from_object("hmalite.config.HmaLiteDevConfig")
-app.config.from_envvar("HMALITE_CONFIG_FILE", silent=True)
+    config_cls = HmaLiteDevConfig
+better_config = config_cls.init_with_environ()
+app.config.from_object(better_config)
+
+if better_config.DEBUG:
+    app.logger.info("Config Values:")
+    for name in config_cls._fields:
+        app.logger.info("%s = %s", name, app.config[name])
+
+
+####################### INDEX HACKS #######################
+# This doesn't really behave the way you might think -
+# instances of flask are destroyed and created all the time/thread
+# local-d, so you'll get a bunch of copies of this(?)
+# Someday we'll have to figure out the correct way to stage it
+_INDEX = None
```

Comment:
Yeah not ideal, I wonder if something like a global 'session' or some kind of cache could work, probably not easily between different API requests...

Reply:
AFAIK, flask application contexts are created and discarded at every request. [1]. Their recommendation for storing data across requests:

> Use the session or a database to store data across requests.

And before you say "session", even that is implemented using memcached/redis. This has always puzzled me about python/php etc. Because of the request-first lifecycle management, there is no way to share state across requests without serialization / deserialization overhead.

Short of implementing a sidecar pattern [2], I do not see a way out of this.

In the mean time, can we `timeit` [3] the initialization and keep logging?

1: https://flask.palletsprojects.com/en/1.1.x/appcontext/#storing-data
2: https://www.oreilly.com/library/view/designing-distributed-systems/9781491983638/ch02.html
3: https://docs.python.org/3/library/timeit.html

Reply:
If this is crucial, I'd second that flask is not the appropriate tech for it. Event Driven systems could be more appropriate. eg. https://www.twistedmatrix.com/trac/

Twisted has long lived resources and they display it in its full-glory on their home page. We might want to give that a shot.

Reply:
1. Application contexts are discarded every requests, but the memory is not cleared. This trick works because it's in the outer scope of memory
2. Re: timit - A few seconds at the larger size, but it will only load once each thread.
3. Re: Twisted - I don't think we should invest in hmalite much more than has been done. Having dug into it a bit, I think lambda is probably good enough, and would want to see measurements before we started changing stuff.